src/main/java/com/example/onlinebookstore/service/order/OrderServiceImpl.java

```diff
@@ -0,0 +1,126 @@
+        return orderItems.stream().map(orderItemMapper::toDto).toList();
+    }
+
+    @Override
+    public OrderItemResponseDto getOrderItem(Long orderId, Long itemId) {
+        OrderItem orderItem = orderItemRepository.findByOrderIdAndId(orderId, itemId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Cant find order item with id " + itemId + " and order id " + orderId));
+        return orderItemMapper.toDto(orderItem);
+    }
+
+    @Override
+    @Transactional
+    public OrderResponseDto updateStatus(Long id, PatchOrderRequestDto requestDto) {
+        Order order = orderRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find order with id " + id));
+        order.setStatus(Order.Status.valueOf(requestDto.getStatus()));
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    @Transactional
+    public OrderResponseDto placeOrder(Long userId, PostOrderRequestDto requestDto) {
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find user with id " + userId));
+
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find shopping cart with id " + userId));
+
+        BigDecimal total = countTotal(shoppingCart);
+
+        Order order = new Order();
+        order.setUser(user);
+        order.setStatus(Order.Status.PENDING);
+        order.setTotal(total);
+        order.setShippingAddress(requestDto.getShippingAddress());
+        orderRepository.save(order);
+
+        Set<OrderItem> orderItems = createOrderItems(shoppingCart, order);
+        orderItemRepository.saveAll(orderItems);
+
+        order.setOrderItems(orderItems);
+        orderRepository.save(order);
+
+        cartItemRepository.deleteAll(shoppingCart.getCartItems());
+        return orderMapper.toDto(order);
+    }
+
+    private BigDecimal countTotal(ShoppingCart shoppingCart) {
+        BigDecimal total = BigDecimal.ZERO;
+        for (CartItem cartItem : shoppingCart.getCartItems()) {
+            BigDecimal cartItemPrice = cartItem.getBook().getPrice()
+                    .multiply(BigDecimal.valueOf(cartItem.getQuantity()));
+            total = total.add(cartItemPrice);
+        }
+        return total;
+    }
+
+    private Set<OrderItem> createOrderItems(ShoppingCart shoppingCart, Order order) {
+        Set<OrderItem> orderItems = new HashSet<>();
+
+        for (CartItem cartItem : shoppingCart.getCartItems()) {
+            Book book = cartItem.getBook();
+
+            OrderItem orderItem = new OrderItem();
+            orderItem.setOrder(order);
+            orderItem.setBook(book);
+            orderItem.setQuantity(cartItem.getQuantity());
```

Comment:
Please verify that if you add order Items with the same quantities they are both added to HashSet. Use case: add 2 items with the same quantity and see that there are 2 items in the set.